### PR TITLE
Jet ID for UL

### PR DESCRIPTION
**Pull request by ahinzmann**
#### PR description:

This adds the jet ID selector tuned on UL data for AK4 jets and updates DQM to use this definition to define "Cleaned" histograms.

Also, the JetID is now also applied to /DiJet/* histograms for RECO jets in DQM, which was previously only applied for MiniAOD jets in DQM.

Presentation of the CHS criteria for UL17+18: https://indico.cern.ch/event/928283/contributions/3917588/
Presentation of the PUPPI criteria for UL17: https://indico.cern.ch/event/937597/contributions/3940302/

This functor is not used in AOD/MiniAOD, but only in NanoAOD. Thus AOD/MiniAOD content are not expected to change but NanoAOD content can change.

The functor is used to define "Cleaned" DQM histograms, which are expected to change.

#### PR validation:

runTheMatrix.py -l limited

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Plan to backport to 10_6_X and integrate in UL-reMiniAOD.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
@ahinzmann please forgive me if my understanding from the PR description is not correct:
- I understand the new AK4 definition is tighter than the previous since entires in MEs get reduced from reference
- There are AK8 folders being modified, not sure if the change should be afecting only AK4 ones
- The description mentions the "AOD/MiniAOD are not expected to change" but there seems to be changes in the MiniAOD_over_RECO folder
In a nutshell, can you check if all the changes are expected:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-07-10-1100+7ef228/37633/dqm-histo-comparison-summary.html

File: PhysicsTools/SelectorUtils/python/pfJetIDSelector_cfi.py [hunk 1]
```diff
@@ -2,6 +2,6 @@
 
 
 pfJetIDSelector = cms.PSet(
-        version = cms.string('WINTER17'),
+        version = cms.string('RUNIIULCHS'),
```
Comment (perrotta) [inline PhysicsTools/SelectorUtils/python/pfJetIDSelector_cfi.py:5]: Shouldn't this be "RUN2..." instead of "RUNII..."?

**Comment by perrotta** [comment]
@jfernan2 , as far as I can see in DQM' jetAnalyzer_cfi.py the defaults are changed from "WINTER16" to "RUN2ULCHS" (for AK4PF*, AK4CHS* jets) and from "WINTER17PUPPI" to "RUN2ULPUPPI" (for Ak8PFPUPPICleaned and Ak4PFCHSPuppiCleaned jets)

**Comment by perrotta** [comment]
@ahinzmann please clarify on the status of this PR
Thank you.

**Comment by ahinzmann** [comment]
> @ahinzmann please forgive me if my understanding from the PR description is not correct:
> 
>     * I understand the new AK4 definition is tighter than the previous since entires in MEs get reduced from reference
> 
>     * There are AK8 folders being modified, not sure if the change should be afecting only AK4 ones
AK8 jets use the AK4 jet ID in DQM, I think.

>     * The description mentions the "AOD/MiniAOD are not expected to change" but there seems to be changes in the MiniAOD_over_RECO folder
The statement refers to no change in the actual content of AOD and MiniAOD, i.e. no changes found in RECO comparisons. The observed changes are due applying a new ID version in DQM when making "Cleaned" plots, so I'd say this is indeed expected.

>       In a nutshell, can you check if all the changes are expected:
>       https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-07-10-1100+7ef228/37633/dqm-histo-comparison-summary.html

**Comment by ahinzmann** [comment]
The DQM comparison plots were actually deleted before I managed to look at them. I will check them again after the next round of tests.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
+1
- Two new jetID selections  (RUN2ULCHS and RUN2ULPUPPI) are defined, to be used in UL analyses and possibly UL nanoAOD production 
- In pfJetIDSelector_cfi.py the default for pfJetIDSelector is changed from 'WINTER17' to 'RUNIIULCHS', following the corresponding change in default of the PFJetIDSelectionFunctor
  - Besides a few test scripts, the pfJetIDSelector is only included in PhysicsTools/SelectorUtils/python/wplusjetsAnalysis_cfi.py (never linked in the standard workflows)
- Jenkins tests pass and show no difference in reco or mini/nanoAOD outputs, as expected for the standard workflows
- Jenkins tests show differences in DQM outputs, which are supposedly expected  because in DQM' jetAnalyzer_cfi.py the default selections are changed from "WINTER16" to "RUN2ULCHS" (for AK4PF*, AK4CHS* jets) and from "WINTER17PUPPI" to "RUN2ULPUPPI" (for Ak8PFPUPPICleaned and Ak4PFCHSPuppiCleaned jets)

**Comment by jfernan2** [comment]
+1

**Comment by ahinzmann** [comment]
Most of the jet ID DQM histograms show only small changes, however, there are a few that are not filled with any event for the pr, e.g.:

https://tinyurl.com/y3rrzmzt

https://tinyurl.com/yy6hxxk4

https://tinyurl.com/y2yzozls

https://tinyurl.com/yy6hxxk4

Is this a feature of these particular workflows or of this PR?

For most other workflows the same histograms are filled as expected.

**Comment by perrotta** [comment]
Removed my "+1" (it was too quick)

@ahinzmann , I think that the new JetIDVersion's must be considered inside DQMOffline/JetMET/src/JetAnalyzer.cc

**Comment by perrotta** [comment]
please test

**Comment by ahinzmann** [comment]
Ok, now the emtpy histograms are gone and differences are smaller. One can see a reduction of events from the tighter jet ID.

One thing that is not obvious to me are the MiniAOD_over_RECO plots in DQM. They are not unity anymore, e.g. in:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-07-21-2300+7ef228/37909/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/JetMET__Jet_MiniAOD_over_RECO_DiJet_NeutralConstituentsFraction_MiniAOD_over_RECO.png
Seems like a different jet ID selection is applied for such plots on MiniAOD and RECO. This is probably not desirable as it make it impossible to spot differences between MiniAOD and RECO. Needs further investigation.

**Comment by perrotta** [comment]
Some lower efficiency of the new ID's at intermediate pt (this is wf 136.731 from the automatic DQM comparisons)

image

And here cleaned slimmed jets, also from  wf 136.731

image

image

image

Differences are much less significant in PUPPI jets (the other modified ID in the DQM plots)

Please confirm that this corresponds to what expected from the new selections

**Comment by ahinzmann** [comment]
I believe the jet ID pass fraction vs. pT is as expected.
In JetHT (i.e. real jet enriched), one observes >99% efficiency:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-07-21-2300+7ef228/37909/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017/JetMET__Jet_CleanedslimmedJets_JetIDPassFractionVSpt.png
image
In SinglePhoton (i.e. fake jet enriched), one observes a lowering in efficiency:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-07-21-2300+7ef228/37909/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/JetMET__Jet_CleanedslimmedJets_JetIDPassFractionVSpt.png
image

The only puzzle are the MiniAOD over RECO plots. Maybe @iatakisi knows more?

**Comment by ahinzmann** [comment]
The JetID is now also applied to /DiJet/* histograms for RECO jets in DQM, which was previously only applied for MiniAOD jets in DQM. This should fix the MiniAOD_over_RECO plots.

**Comment by perrotta** [comment]
please test

**Comment by ahinzmann** [comment]
Looks like expected now. MiniAOD_over_RECO becomes unity everywhere.

**Comment by perrotta** [comment]
+1
- (Now that also DQM looks like having been fixed, see https://github.com/cms-sw/cmssw/pull/30638#issuecomment-662859310, I can repeat the reco signature already given in https://github.com/cms-sw/cmssw/pull/30638#issuecomment-662248545)
- Two new jetID selections (RUN2ULCHS and RUN2ULPUPPI) are defined, to be used in UL analyses and possibly UL nanoAOD production
- In pfJetIDSelector_cfi.py the default for pfJetIDSelector is changed from 'WINTER17' to 'RUNIIULCHS', following the corresponding change in default of the PFJetIDSelectionFunctor
  - Besides a few test scripts, the pfJetIDSelector is only included in PhysicsTools/SelectorUtils/python/wplusjetsAnalysis_cfi.py (never linked in the standard workflows)
- Jenkins tests pass and show no difference in reco or mini/nanoAOD outputs, as expected for the standard workflows
- Jenkins tests show differences in DQM outputs, which are supposedly expected because in DQM' jetAnalyzer_cfi.py the default selections are changed from "WINTER16" to "RUN2ULCHS" (for AK4PF*, AK4CHS* jets) and from "WINTER17PUPPI" to "RUN2ULPUPPI" (for Ak8PFPUPPICleaned and Ak4PFCHSPuppiCleaned jets)

**Comment by jfernan2** [comment]
+1

**Comment by santocch** [comment]
+1

**Comment by silviodonato** [comment]
+1

File: DQMOffline/JetMET/python/jetAnalyzer_cfi.py [hunk 2]
```diff
@@ -106,8 +106,8 @@
     #for PFJets: LOOSE,TIGHT
     JetIDQuality               = cms.string("LOOSE"),
```
Comment (ahinzmann) [inline DQMOffline/JetMET/python/jetAnalyzer_cfi.py:107]: this default should be put to TIGHT as default
